src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ApplicationAssembliesProvider.cs

```diff
@@ -0,0 +1,283 @@
+
+            assemblyItems = assemblyItems
+                .OrderBy(item => item.Assembly == entryAssembly ? 0 : 1)
+                .ThenBy(item => item.Assembly.FullName, StringComparer.Ordinal);
+
+            foreach (var item in assemblyItems)
+            {
+                yield return item.Assembly;
+
+                foreach (var associatedAssembly in item.RelatedAssemblies.Distinct().OrderBy(assembly => assembly.FullName, StringComparer.Ordinal))
+                {
+                    yield return associatedAssembly;
+                }
+            }
+        }
+
+        protected virtual DependencyContext LoadDependencyContext(Assembly assembly) => DependencyContext.Load(assembly);
+
+        private List<AssemblyItem> ResolveFromDependencyContext(DependencyContext dependencyContext)
+        {
+            var assemblyItems = new List<AssemblyItem>();
+            var relatedAssemblies = new Dictionary<Assembly, AssemblyItem>();
+
+            var candidateAssemblies = GetCandidateLibraries(dependencyContext)
+                .SelectMany(library => GetLibraryAssemblies(dependencyContext, library));
+
+            foreach (var assembly in candidateAssemblies)
+            {
+                var assemblyItem = GetAssemblyItem(assembly);
+                assemblyItems.Add(assemblyItem);
+
+                foreach (var relatedAssembly in assemblyItem.RelatedAssemblies)
+                {
+                    if (relatedAssemblies.TryGetValue(relatedAssembly, out var otherEntry))
+                    {
+                        var message = string.Join(
+                            Environment.NewLine,
+                            Resources.FormatApplicationAssembliesProvider_DuplicateRelatedAssembly(relatedAssembly.FullName),
+                            otherEntry.Assembly.FullName,
+                            assembly.FullName);
+
+                        throw new InvalidOperationException(message);
+                    }
+
+                    relatedAssemblies.Add(relatedAssembly, assemblyItem);
+                }
+            }
+
+            // Remove any top level assemblies that appear as an associated assembly.
+            assemblyItems.RemoveAll(item => relatedAssemblies.ContainsKey(item.Assembly));
+
+            return assemblyItems;
+        }
+
+        protected virtual IEnumerable<Assembly> GetLibraryAssemblies(DependencyContext dependencyContext, RuntimeLibrary runtimeLibrary)
+        {
+            foreach (var assemblyName in runtimeLibrary.GetDefaultAssemblyNames(dependencyContext))
+            {
+                var assembly = Assembly.Load(assemblyName);
+                yield return assembly;
+            }
+        }
+
+        protected virtual IReadOnlyList<Assembly> GetRelatedAssemblies(Assembly assembly)
+        {
+            // Do not require related assemblies to be available in the default code path.
+            return RelatedAssemblyAttribute.GetRelatedAssemblies(assembly, throwOnError: false);
```

Comment:
Why is this? This makes it incredibly hard to debug if there is no related assembly in place due to something. Whats the trade-off?

Reply:
I'll file an issue in Razor tracking this for preview2. Maybe there's a simple way to determine out what context (`Build` vs `Publish`) you're running in. If we can resolve that, we can remove this parameter.

Reply:
I might be able to help you with this. We’ll chat about this when I’m in the office